### PR TITLE
Add tests for truncated and symlinked files

### PR DESCRIPTION
**Pull request by kvch**
## What does this PR do?

Add support for better support truncated files with symlinks and adds migrates related tests from `test_harvester.py`.

## Why is it important?

Increase coverage.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
~~- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

### CODE REVIEW

File: filebeat/input/filestream/fswatch.go [hunk 1]
```diff
@@ -232,20 +232,20 @@ func (w *fileWatcher) GetFiles() map[string]os.FileInfo {
 	return w.scanner.GetFiles()
 }
 
-type fileScannerConfig struct {
+type FileScannerConfig struct {
```
Comment (urso) [inline filebeat/input/filestream/fswatch.go:235]: Why do we need to export that symbol?

Comment (kvch) [reply filebeat/input/filestream/fswatch.go]: Not needed, removed.

File: filebeat/input/filestream/input_integration_test.go [hunk 2]
```diff
@@ -736,3 +737,174 @@ func TestFilestreamTruncateBlockedOutput(t *testing.T) {
...
+// test_symlinks_enabled from test_harvester.py
+func TestFilestreamSymlinksEnabled(t *testing.T) {
+	env := newInputTestingEnvironment(t)
+
+	testlogName := "test.log"
+	symlinkName := "test.log.symlink"
+	inp := env.mustCreateInput(map[string]interface{}{
+		"paths": []string{
+			env.abspath(testlogName),
+			env.abspath(symlinkName),
```
Comment (urso) [inline filebeat/input/filestream/input_integration_test.go:750]: The test might not capture everything going on here. The configuration looks like the original log file is collected and that we collect via symlinks. Is this on purpose? If so, what is the purpose of this test?

**Comment by kvch** [comment]
jenkins run tests

File: filebeat/input/filestream/input_integration_test.go (hunk 2 above)
Comment (kvch) [reply filebeat/input/filestream/input_integration_test.go:749]: I copied the configuration from `paths` from the original test incorrectly. I have also made the same mistake in the other 3 tests. I have fixed the paths and the corresponding assertions.

File: filebeat/input/filestream/input_integration_test.go [hunk 3]
```diff
@@ -736,3 +737,166 @@ func TestFilestreamTruncateBlockedOutput(t *testing.T) {
...
+	env.startInput(ctx, inp)
+
+	env.waitUntilEventCount(1)
+
+	expectedOffset := len(commonLine) + 2
+	env.requireOffsetInRegistry(firstTestlogName, expectedOffset)
+
+	// rotate symlink
+	env.mustRemoveFile(symlinkName)
+	env.mustSymlink(secondTestlogName, symlinkName)
```
Comment (urso) [inline filebeat/input/filestream/input_integration_test.go:804]: Interesting. I would not expect the symlink to be removed when rotating it. Just change the location it points to. What would happen in that case? What would happen if the new log file is larger then the old one when we update the link?

File: filebeat/input/filestream/input_integration_test.go [hunk 4]
```diff
@@ -736,3 +737,166 @@ func TestFilestreamTruncateBlockedOutput(t *testing.T) {
...
+
+	env.requireOffsetInRegistry(testlogName, len(line))
+
+	// remove symlink
+	env.mustRemoveFile(symlinkName)
+
+	env.mustAppendLinesToFile(testlogName, line)
+
+	env.waitUntilEventCount(2)
+	env.requireOffsetInRegistry(testlogName, 2*len(line))
```
Comment (urso) [inline filebeat/input/filestream/input_integration_test.go:853]: What would happen if we have `close...removed: true`? Will remove act on the symlink state, or on the resolved file?

What happens on `close...renamed`?

File: filebeat/input/filestream/input_integration_test.go (hunk 3 above)
Comment (kvch) [reply filebeat/input/filestream/input_integration_test.go:804]: I have just ported blindly the test from the original Python one. There should not be any change in behaviour of Filebeat if the symlink is renamed.

If the log file is bigger than the previous, Filebeat cannot detect truncation regardless of what we do with the file symlink, rename, remove and create again. That is why we need the special copytruncate prospector.

File: filebeat/input/filestream/input_integration_test.go (hunk 4 above)
Comment (kvch) [reply filebeat/input/filestream/input_integration_test.go:853]: Filebeat cares about the original file, not the symlink. The setting has no impact on the behaviour of the reader if the symlink is deleted. I am not sure why the original test contains it.

File: filebeat/input/filestream/input_integration_test.go (hunk 3 above)
Comment (urso) [reply filebeat/input/filestream/input_integration_test.go:804]: I see. So we do track the inode of the symlink file, and not the inode of the target file.

I guess we need to rethink symlink support to some extent. Detecting changes on a symlink should not only depend on the file size. Have we also considered the case of a symlink becoming an actual file (or the other way around)?

File: filebeat/input/filestream/input_integration_test.go [hunk 5]
```diff
@@ -736,3 +737,166 @@ func TestFilestreamTruncateBlockedOutput(t *testing.T) {
...
+
+	ctx, cancelInput := context.WithCancel(context.Background())
+	env.startInput(ctx, inp)
+
+	env.waitUntilEventCount(1)
+
+	env.requireOffsetInRegistry(testlogName, len(line))
+
+	// remove symlink
+	env.mustRemoveFile(symlinkName)
```
Comment (urso) [inline filebeat/input/filestream/input_integration_test.go:848]: What happens if I don't remove/update the symlink, but the file the symlink points to? Do we even detect that case?

File: filebeat/input/filestream/input_integration_test.go [hunk 6]
```diff
@@ -736,3 +737,166 @@ func TestFilestreamTruncateBlockedOutput(t *testing.T) {
...
+	env.startInput(ctx, inp)
+
+	env.waitUntilEventCount(3)
+
+	env.requireOffsetInRegistry(testlogName, len(lines))
+
+	// remove symlink
+	env.mustRemoveFile(symlinkName)
+	env.mustTruncateFile(testlogName, 0)
+	env.waitUntilOffsetInRegistry(testlogName, 0)
```
Comment (urso) [inline filebeat/input/filestream/input_integration_test.go:890]: Why does the test remove the symlink?

File: filebeat/input/filestream/input_integration_test.go (hunk 5 above)
Comment (kvch) [reply filebeat/input/filestream/input_integration_test.go:848]: Yes, the reader opens the original file and tracks everything the same way as done in ordinary files.

File: filebeat/input/filestream/input_integration_test.go (hunk 3 above)
Comment (kvch) [reply filebeat/input/filestream/input_integration_test.go:804]: Is it OK if I add a test for the scenario you had described in a follow up PR? You are raising valid points. However, we should decide what we want to tackle in the default prospector and what we want to cover in the copytruncate prospector. I would rather give us a bit more time for that and address your concerns in a new PR. WDYT?

**Comment by urso** [comment]
ARM tests fail because we are hitting docker registry limits. Are we missing some images in our caches?

**Review by urso — APPROVED** [review]

**Pull request merged**